plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergBucketFunction.java

```diff
@@ -0,0 +1,140 @@
+
+import io.trino.plugin.iceberg.PartitionTransforms.ColumnTransform;
+import io.trino.plugin.iceberg.PartitionTransforms.ValueTransform;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.connector.BucketFunction;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.TypeOperators;
+import org.apache.iceberg.PartitionSpec;
+
+import java.lang.invoke.MethodHandle;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.iceberg.PartitionTransforms.getColumnTransform;
+import static io.trino.spi.function.InvocationConvention.InvocationArgumentConvention.NEVER_NULL;
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.trino.spi.function.InvocationConvention.simpleConvention;
+import static io.trino.spi.type.TypeUtils.NULL_HASH_CODE;
+import static java.util.Objects.requireNonNull;
+
+public class IcebergBucketFunction
+        implements BucketFunction
+{
+    private final int bucketCount;
+
+    private final List<PartitionColumn> partitionColumns;
+    private final List<MethodHandle> hashCodeInvokers;
+
+    public IcebergBucketFunction(
+            TypeOperators typeOperators,
+            PartitionSpec partitionSpec,
+            List<IcebergColumnHandle> partitioningColumns,
+            int bucketCount)
+    {
+        requireNonNull(partitionSpec, "partitionSpec is null");
+        checkArgument(!partitionSpec.isUnpartitioned(), "empty partitionSpec");
+        checkArgument(bucketCount > 0, "Invalid bucketCount: %s", bucketCount);
+        this.bucketCount = bucketCount;
+
+        Map<Integer, Integer> fieldIdToInputChannel = new HashMap<>();
+        for (int i = 0; i < partitioningColumns.size(); i++) {
+            fieldIdToInputChannel.put(partitioningColumns.get(i).getId(), i);
+        }
+        partitionColumns = partitionSpec.fields().stream()
+                .map(field -> {
+                    Integer channel = fieldIdToInputChannel.get(field.sourceId());
+                    checkArgument(channel != null, "partition field not found: %s", field);
+                    Type inputType = partitioningColumns.get(channel).getType();
+                    ColumnTransform transform = getColumnTransform(field, inputType);
+                    return new PartitionColumn(channel, transform.getValueTransform(), transform.getType());
+                })
+                .collect(toImmutableList());
+        hashCodeInvokers = partitionColumns.stream()
+                .map(PartitionColumn::getResultType)
+                .map(type -> typeOperators.getHashCodeOperator(type, simpleConvention(FAIL_ON_NULL, NEVER_NULL)))
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public int getBucket(Page page, int position)
```

Comment:
I wonder if the API should be `int[] getBucket(Page page, SelectedPositions positions)` instead.
This one doesn't look great for vectorization. If we had a list of positions instead, maybe we could also take advantage of RLE or Dictionary block to reduce hash computations.
cc: @sopel39 @skrzypo987

Reply:
Yup. We can consider vectorized API. Yet I think it's out of scope here. BTW We can always isolate partition generation code, which should inline whatever bucket function is used.

Reply:
> I wonder if the API should be int[] getBucket(Page page, SelectedPositions positions) instead.

or all of positions. idk. 

> Yet I think it's out of scope here.

yes, absolutely.

-> https://github.com/trinodb/trino/issues/9848